core/domain/state_domain.py

```diff
@@ -916,8 +940,1085 @@ def get_all_outcomes(self) -> List[Outcome]:
+        setting is turned on
+        - `IsExactlyEqualTo` rule should not have integral value when
+        `allow non zero integers` setting is off
+        - All rules should have solutions that is not subset of
+        previous rules' solutions
+        - `HasDenominatorEqualTo` rule should not have value less than 0
+        - `HasFractionalPartExactlyEqualTo` rule should always come before
+        `HasDenominatorEqualTo` rule where the fractional denominator is
+        equal to `HasDenominatorEqualTo` rule value
+
+        Raises:
+            ValidationError. Raises a validation error when the above
+                mentioned checks fails.
+        """
+        ranges: List[InteractionInstance.RangeVariable] = []
+        matched_denominator_list: List[
+            InteractionInstance.MatchedDenominator] = []
+        rule_spec_till_now: List[state_domain.RuleSpecDict] = []
+        inputs_with_whole_nums = [
+            'HasDenominatorEqualTo',
+            'HasNumeratorEqualTo',
+            'HasIntegerPartEqualTo',
+            'HasNoFractionalPart'
+        ]
+        lower_infinity = float('-inf')
+        upper_infinity = float('inf')
+        allow_non_zero_integ_part = (
+            self.customization_args[
+                'allowNonzeroIntegerPart'].value)
+        allow_imp_frac = (
+            self.customization_args[
+                'allowImproperFraction'].value)
+        require_simple_form = (
+            self.customization_args[
+                'requireSimplestForm'].value)
+        for ans_group_index, answer_group in enumerate(self.answer_groups):
+            for rule_spec_index, rule_spec in enumerate(
+                answer_group.rule_specs):
+                # Rule should not be duplicate.
+                if rule_spec.to_dict() in rule_spec_till_now:
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer group '
+                        f'\'{ans_group_index}\' of FractionInput '
+                        f'interaction is already present.'
+                    )
+                else:
+                    rule_spec_till_now.append(rule_spec.to_dict())
+
+                if rule_spec.rule_type not in inputs_with_whole_nums:
+                    num = rule_spec.inputs['f']['numerator']
+                    den = rule_spec.inputs['f']['denominator']
+                    whole = rule_spec.inputs['f']['wholeNumber']
+
+                    # Denominator should be greater than 0.
+                    if whole == 0:
+                        if den <= 0:
+                            raise utils.ValidationError(
+                                f'The rule \'{rule_spec_index}\' of answer '
+                                f'group \'{ans_group_index}\' has '
+                                f'denominator less than or equal to zero '
+                                f'in FractionInput interaction.'
+                            )
```

Comment:
I'm thinking about this again and don't think that denominator < 0 is an issue. Having a negative denominator seems okay to me. The issue is just denominator = 0. Can we change this? Also don't understand why this needs to be wrapped in the `if whole == 0` if you could explain

Reply:
So previously when I checked, if we only enter an whole number value then the numerator and denominator both would be `0` so just to check for the fraction part I did this. I have updated it now though.
Regarding `<0` part - This check is also present in the controller layer stating that the minimum value should be `1` and also present on our [validation sheet](https://docs.google.com/spreadsheets/d/1kNLF-UkS_Qk0igNCGwLjJXM0ByHalqli41GPvWF43PE/edit#gid=290682831). Please let me know if we plan to update this.
Thanks for pointing it out

Reply:
Ah I see that makes sense. Thanks for explaining!